server/datastore/mysql/software_installers.go

```diff
@@ -0,0 +1,136 @@
+package mysql
+
+import (
+	"context"
+	"database/sql"
+
+	"github.com/fleetdm/fleet/v4/server/contexts/ctxerr"
+	"github.com/fleetdm/fleet/v4/server/fleet"
+	"github.com/jmoiron/sqlx"
+)
+
+func (ds *Datastore) MatchOrCreateSoftwareInstaller(ctx context.Context, payload *fleet.UploadSoftwareInstallerPayload) (uint, error) {
+	titleID, err := ds.getOrGenerateSoftwareInstallerTitleID(ctx, payload.Title, payload.Source)
+	if err != nil {
+		return 0, err
+	}
+
+	installScriptID, err := ds.getOrGenerateScriptContentsID(ctx, payload.InstallScript)
+	if err != nil {
+		return 0, err
+	}
+
+	var postInstallScriptID *uint
+	if payload.PostInstallScript != "" {
+		sid, err := ds.getOrGenerateScriptContentsID(ctx, payload.PostInstallScript)
+		if err != nil {
+			return 0, err
+		}
+		postInstallScriptID = &sid
+	}
+
+	var tid uint
+	if payload.TeamID != nil {
+		tid = *payload.TeamID
+	}
+
+	stmt := `
+INSERT INTO software_installers (
+	team_id,
+	global_or_team_id, 
+	title_id, 
+	storage_id,
+	filename, 
+	version,
+	install_script_content_id, 
+	pre_install_query, 
+	post_install_script_content_id
+) VALUES (?, ?, ?, ?, ?, ?, ?, ?, ?)`
+
+	args := []interface{}{
+		payload.TeamID,
+		tid,
+		titleID,
+		payload.StorageID,
+		payload.Filename,
+		payload.Version,
+		installScriptID,
+		payload.PreInstallQuery,
+		postInstallScriptID,
+	}
+
+	res, err := ds.writer(ctx).ExecContext(ctx, stmt, args...)
+	if err != nil {
+		if isDuplicate(err) {
+			// already exists for this team/no team
+			err = alreadyExists("SoftwareInstaller", payload.Title)
+		}
+		return 0, ctxerr.Wrap(ctx, err, "insert software installer")
+	}
+
+	id, _ := res.LastInsertId()
+
+	return uint(id), nil
+}
+
+func (ds *Datastore) getOrGenerateSoftwareInstallerTitleID(ctx context.Context, name, source string) (uint, error) {
+	titleID, err := ds.optimisticGetOrInsert(ctx,
+		&parameterizedStmt{
+			Statement: `SELECT id FROM software_titles WHERE name = ? AND source = ? AND browser = ''`,
```

Comment:
based on the data from Martin, I think we'll want to make this more flexible. Eg: also look at bundle identifiers for matches.

we can tweak it later.